File: cs/codewars/find_the_parity_outlier.cs
// https://www.codewars.com/kata/find-the-parity-outlier

// Attempt 1:
// Probably faster

using System.Linq;

public class Kata {
  public static int Find(int[] xs) {
    var findOdd = IsEven(xs[0]) && IsEven(xs[1]);
    return xs.First(x => findOdd ^ IsEven(x));
  }

  private static bool IsEven(int x) => x % 2 == 0;
}

// Attempt 2:
// Probably slower, but "fancier"

using System.Linq;

public class Kata {
  public static int Find(int[] xs) {
    var evens = xs.Where(IsEven);
    var odds  = xs.Where(x => !IsEven(x));
    return evens.Skip(1).Any() ? odds.First() : evens.First();
  }

  private static bool IsEven(int x) => x % 2 == 0;
}

File: hs/codewars/find_the_parity_outlier.hs
-- https://www.codewars.com/kata/find-the-parity-outlier

module Kata where

findOutlier :: [Int] -> Int
findOutlier xs = if (drop 1 evens) == [] then evens !! 0 else odds !! 0
  where
    evens = filter (\x -> x `mod` 2 == 0) xs
    odds  = filter (\x -> x `mod` 2 == 1) xs
